lib/cloud-metadata/callback-coordination.js

```diff
@@ -0,0 +1,145 @@
+'use strict'
+const { EventEmitter } = require('events')
+
+// Indirect usage of the singleton `Agent` to log.
+function getLogger () {
+  return require('../..').logger
+}
+
+/**
+ * Coordinates fetching of metadata from multiple providers
+ *
+ * Implements event based coordination for fetching metadata
+ * from multiple providers.  The first provider to return
+ * a non-error result "wins".  When this happens the CallbackCoordination
+ * object will emit a `result` event.
+ *
+ * If all the metadata providers fail to return a result, then the
+ * object will emit an error event indicating failure to collect metadata
+ * from any event.
+ *
+ * Since a scheduled callback may (accidently) fail to call its own
+ * callback function, the CallbackCoordination object includes its
+ * own timeout timer to avoid deadlock situation.
+ */
+class CallbackCoordination extends EventEmitter {
+  constructor (maxWaitMS = -1) {
+    super()
+    // how many results have we seen
+    this.resultCount = 0
+    this.expectedResults = 0
+    this.errors = []
+    this.scheduled = []
+    this.done = false
+
+    this.timeout = null
+    if (maxWaitMS !== -1) {
+      this.timeout = setTimeout(() => {
+        if (!this.done) {
```

Comment:
I'm going to pass on the `assert` suggestions -- it's good general advice but if we crash from a failed assertion that means we crash a user's/customer's application, and that's typically something we (APM vendors) try to avoid in during in app/process instrumentation. (Be exceptional about not having uncaught exceptions, etc.)

Reply:
Noting that, per conversations ~offline~ elsewhere, we are logging an error here instead of doing a hard assert/exit.